firmware/STN1110.c

```diff
@@ -29,56 +29,102 @@
 {
     sdWrite(&SD2, (uint8_t*)at_cmd, strlen(at_cmd));
-    chThdSleepMilliseconds(100);
+    sdWrite(&SD2, (uint8_t*)"\r", 1);
+    chThdSleepMilliseconds(AT_COMMAND_DELAY);
+}
+
+/* Send an AT command with a single numberic parameter value */
+static void _send_at_param(char *at_cmd, int param)
+{
+    sdWrite(&SD2, (uint8_t*)at_cmd, strlen(at_cmd));
+    char param_str[20];
+    modp_itoa10(param, param_str);
+    sdWrite(&SD2, (uint8_t*)param_str, strlen(param_str));
+    sdWrite(&SD2, (uint8_t*)"\r", 1);
+    chThdSleepMilliseconds(AT_COMMAND_DELAY);
 }
 
-void stn1110_reset(uint8_t protocol)
+/* Send the STN1110 command to report the currently detected OBDII protocol */
+static void _send_detect_protocol(void)
 {
-	set_system_initialized(false);
+    /* Wait for STN1110 chip to be ready for command */
+    chThdSleepMilliseconds(AT_COMMAND_DELAY);
+    _send_at("AT DP");
+}
+
+/* Perform a hard reset of the STN1110 */
+void stn1110_reset(enum obdii_protocol protocol, enum obdii_adaptive_timing adaptive_timing, uint8_t obdii_timeout)
+{
+    set_system_initialized(false);
     log_info(LOG_PFX "Reset STN1110 - protocol %i\r\n", protocol);
 
     /* set STN1110 NVM reset to disbled (normal running mode)
      * Use internall pullup resistor to disable NVM
+     * TODO: this will be changed in hardware to just tie it to 3.3v
+     * since we don't really need processor control of this pin
      * */
     palSetPadMode(GPIOA, GPIOB_RESET_NVM_STN1110, PAL_MODE_INPUT_PULLUP);
 
     /* Toggle hard reset Line */
     palSetPadMode(GPIOB, GPIOB_RESET_STN1110, PAL_MODE_OUTPUT_PUSHPULL);
     palClearPad(GPIOB, GPIOB_RESET_STN1110);
-    chThdSleepMilliseconds(10);
+    chThdSleepMilliseconds(RESET_DELAY);
     palSetPad(GPIOB, GPIOB_RESET_STN1110);
-    chThdSleepMilliseconds(1000);
-    log_info(LOG_PFX "after hard reset\r\n");
+    chThdSleepMilliseconds(LONG_DELAY);
+    log_info(LOG_PFX "After hard reset\r\n");
 
-    _send_at("AT E0\r");
+    system_serial_init_SD2(STN1110_INITIAL_BAUD_RATE);
 
-    _send_at("AT SP 0\r");
+    /* switch to the target baud rate */
+    _send_at_param("ST SBR ", STN1110_RUNTIME_BAUD_RATE);
+    system_serial_init_SD2(STN1110_RUNTIME_BAUD_RATE);
 
-    _send_at("AT DPN\r");
+    /* set adaptive timing */
+    _send_at_param("AT AT", adaptive_timing);
 
-    _send_at("ST SBR " STR(STN1110_RUNTIME_BAUD_RATE) "\r");
+    /* set obdii protocol timeout */
+    if (obdii_timeout > 0) {
+        _send_at_param("AT ST ", obdii_timeout);
+    }
 
-    system_serial_init_SD2(STN1110_RUNTIME_BAUD_RATE);
+    /* set protocol */
+    _send_at_param("AT SP ", protocol);
 
-    chThdSleepMilliseconds(1000);
+    /* Disable echo */
+    _send_at("AT E0");
 
+    chThdSleepMilliseconds(LONG_DELAY);
+
+    /* Reset our counters and flags */
+    set_obdii_request_timeout(OBDII_INITIAL_TIMEOUT);
+    set_pid_request_active(false);
+    reset_pid_poll_delay();
+    set_detected_protocol(obdii_protocol_auto);
```

Comment:
Is that a sane default.  Is there ever a case where you may not know the protocol?